qiskit_experiments/calibration/experiments/spectroscopy.py

```diff
@@ -0,0 +1,180 @@
+
+
+class SpectroscopyAnalysis(BaseAnalysis):
+    """Class to analysis a spectroscopy experiment."""
+
+    #pylint: disable=arguments-differ, unused-argument
+    def _run_analysis(
+            self,
+            experiment_data,
+            data_processor=None,
+            meas_level=MeasLevel.CLASSIFIED
+    ) -> Tuple[AnalysisResult, None]:
+        """Analyse the spectroscopy experiment. The analysis will loop over the initial
+        guesses of the full width at half maximum.
+        TODO Missing bounds.
+        TODO Missing user-provided intial guess.
+        """
+
+        # Pick a data processor.
+        if data_processor is None:
+            if meas_level == MeasLevel.CLASSIFIED:
+                data_processor = DataProcessor("counts", [Probability("1")])
+            elif meas_level == MeasLevel.KERNELED:
+                data_processor = DataProcessor("memory", [Svd()])
+            else:
+                raise ValueError("Unsupported measurement level.")
+
+        unit = experiment_data.data[0]["metadata"].get("unit", "Hz")
+
+        y_sigmas = np.array([data_processor(datum) for datum in experiment_data.data])
+        sigmas = y_sigmas[:, 1]
+        ydata = abs(y_sigmas[:, 0])
+        xdata = np.array(datum["metadata"]["xval"] for datum in experiment_data.data)
+
+        offset_guess = np.average(ydata)
+        amp_guess = np.max(ydata)
+        peak_idx = np.argmax(ydata)
+        x0_guess = xdata[peak_idx]
+        gamma_guesses = np.linspace(0, abs(xdata[-1] - xdata[0]), 10)
+
+        best_fit = None
+
+        for gamma_guess in gamma_guesses:
+            fit_result = curve_fit(
+                lambda x, a, g, x0, b: a * ( g**2 / ((x-x0)**2 + g**2)) + b,
+                xdata,
+                np.array(ydata),
+                np.array([amp_guess, gamma_guess, x0_guess, offset_guess]),
+                np.array(sigmas)
+            )
+
+            if not best_fit:
+                best_fit = fit_result
+            else:
+                if fit_result["reduced_chisq"] < best_fit["reduced_chisq"]:
+                    best_fit = fit_result
+
+        analysis_result = AnalysisResult(
+            {
+                "value": best_fit["popt"][2],
+                "stderr": best_fit["popt_err"][2],
+                "unit": unit,
+                "label": "Spectroscopy",
+                "fit": best_fit,
+                "quality": self._fit_quality(
+                    best_fit["popt"], best_fit["popt_err"], best_fit["reduced_chisq"]
+                ),
+            }
+        )
+
+        return analysis_result, None
+
+    @staticmethod
+    def _fit_quality(fit_out, fit_err, reduced_chisq):
+        """TODO"""
+        return ""
+
+
+class Spectroscopy(BaseExperiment):
+    """Class the runs spectroscopy by sweeping the qubit frequency."""
+
+    __analysis_class__ = SpectroscopyAnalysis
+
+    # Supported units for spectroscopy.
+    __units__ = {"Hz": 1.0, "kHz": 1.e3, "MHz": 1.e6, "GHz": 1.e9}
```

Comment:
It's the role of Terra to provide conversion methods between units. It's possibly already somewhere in Terra, and if so, should be used. If not then some day it can be good to move this logic to Terra.

Reply:
I had a look in terra and could not find a unit management system.